lib/grant-types/authorization-code-grant-type.js

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * Module dependencies.
+ */
+
+var AbstractGrantType = require('./abstract-grant-type');
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidGrantError = require('../errors/invalid-grant-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var Promise = require('bluebird');
+var ServerError = require('../errors/server-error');
+var is = require('../validator/is');
+var util = require('util');
+
+/**
+ * Constructor.
+ */
+
+function AuthorizationCodeGrantType(options) {
+  options = options || {};
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getAuthorizationCode()`');
+  }
+
+  if (!options.model.revokeAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `revokeAuthorizationCode()`');
+  }
+
+  if (!options.model.saveToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveToken()`');
+  }
+
+  AbstractGrantType.call(this, options);
+}
+
+/**
+ * Inherit prototype.
+ */
+
+util.inherits(AuthorizationCodeGrantType, AbstractGrantType);
+
+/**
+ * Handle authorization code grant.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+ */
+
+AuthorizationCodeGrantType.prototype.handle = function(request, client) {
```

Comment:
Note that each grant-type exposes a `handle()` function which is called [here](https://github.com/thomseddon/node-oauth2-server/pull/203/files#diff-f3ac454512815b3440a11af47dc3412cR210).

I can certainly rename this to something else, but it would need to be shared with all other grant types, or the underlying architecture itself will need to change.

Any suggestions on method names?